app/admin/tenants/onboarding.py

```diff
@@ -126,94 +124,129 @@ async def onboard_issuer(
         issuer_controller (AcaPyClient): authenticated ACA-Py client for issuer
         endorser_controller (AcaPyClient): authenticated ACA-Py client for endorser
     """
-    # Make sure the issuer has a public did
+
     try:
         issuer_did = await acapy_wallet.get_public_did(controller=issuer_controller)
     except CloudApiException:
         # no public did
+        endorser_did = await acapy_wallet.get_public_did(controller=endorser_controller)
+
+        # Make sure the issuer has a connection with the endorser
+        invitation = await endorser_controller.out_of_band.create_invitation(
+            auto_accept=True,
+            body=InvitationCreateRequest(
+                alias=ACAPY_ENDORSER_ALIAS,
+                handshake_protocols=["https://didcomm.org/didexchange/1.0"],
+                use_public_did=True,
+            ),
+        )
+
+        logger.info(
+            f"Starting webhook listener for connections with wallet id {issuer_wallet_id}"
+        )
+
+        wait_for_event, _ = await start_listener(
+            topic="connections", wallet_id=issuer_wallet_id
+        )
+        endorser_wait_for_connection, _ = await start_listener(
+            topic="connections", wallet_id="admin"
+        )
+
+        endorser_wait_for_transaction, _ = await start_listener(
+            topic="endorsements", wallet_id="admin"
+        )
+
+        logger.debug("Receiving connection invitation")
+
+        # FIXME: make sure the connection with this alias doesn't exist yet
+        # Or does use_existing_connection take care of this?
+        connection_record = await issuer_controller.out_of_band.receive_invitation(
+            auto_accept=True,
+            use_existing_connection=True,
+            body=invitation.invitation,
+            alias=ACAPY_ENDORSER_ALIAS,
+        )
+
+        logger.debug(
+            f"Waiting for connection with id {connection_record.connection_id} to be completed"
+        )
+
+        # Wait for connection to be completed before continuing
+        try:
+            await wait_for_event(
+                filter_map={
+                    "connection_id": connection_record.connection_id,
+                    "state": "completed",
+                }
+            )
```

Comment:
This is a bit useless I think if we have the completed listener for the endorser, and we aren't using the event data. I think we can just remove the connection completed listener for the new tenant completely?

Reply:
ah yeah I think you're correct. I initially thought we'd need to wit for connection completed but as you say waiting for endorser connection to be completed should suffice